### DISCUSSION TITLE
How to store jobs created to find them in a easy way ?

### DISCUSSION THREAD

**Discussion by ddavid67** [taskforcesh/bullmq#388] [Q&A]
Hello

When I create a new job, It is often related to something in my database (a user for example).
What is the best way for the user to find the job he created ? 
It seems I need to store the job id somewhere in my db, and add my context data with it but I am not sure because :

Why is the id field  "optionnal" ? 
It seems to be the most important part of each job so I don't understand why it could be "undefined" .
Is it really unique ? like a redis increment ?
Is it scoped to each queue ? or global ?

**Comment by manast** [comment]
A unique jobid is generated automatically which is in fact a counter, however you can pass your own jobId. However keep in mind that if there is already a job in the queue with the same id then the new job will not be added (this is useful for idempotence and debounce scenarios).

**Comment by ddavid67** [comment]
This doesn't explain why the id field is typed as optionnal



```js
const job= await Job.fromId(queue, id)
  if (job) {
    job.id // <--- typed as "id?"
  }
```

why ?
in what scenario could a job be created without an id?
even if I just successfully resolved a job with the id I just provided...

**Comment by manast** [comment]
@daeschlerdavid All jobs have unique ids, either you let bullmq generate the id or you provide your own. That the id is optional is an implementation detail, you should never need to instantiate a Job manually but indirectly using static functions such as Job.fromId. It is probably possible to make the id non-optional, I am open to suggestions on how to improve the current design.

**Comment by ddavid67** [comment]
Ok. I just assumed that because the id is typed as optionnal that this field is not a reliable way to store/retrieve jobs.

Currently I am using typescript non-null assertion operator to make my code compile with typescript, like this

job.id!
